libr/debug/dsession.c

```diff
@@ -231,3 +235,375 @@ R_API bool r_debug_session_add_mem_change(RDebugSession *session, ut64 addr, ut8
+	pj_kn (j, "size", vmem->len);
+	pj_ka (j, "a");
+	r_vector_foreach (vmem, mem) {
+		pj_o (j);
+		pj_kN (j, "cnum", mem->cnum);
+		pj_kn (j, "data", mem->data);
+		pj_end (j);
+	}
+	pj_end (j);
+	pj_end (j);
+	return true;
+}
+
+// memorylist=<addr>,<addr>,...
+static bool save_memorylist(void *db, const ut64 k, const void *v) {
+	sdb_array_add (db, "memorylist", sdb_fmt ("0x%"PFMT64x, k), 0);
+	return true;
+}
+
+static bool save_memory(Sdb *db, HtUP *memory) {
+	PJ *j = pj_new ();
+	if (!j) {
+		return false;
+	}
+	pj_o (j);
+	ht_up_foreach (memory, save_memory_cb, j);
+	pj_end (j);
+	sdb_set (db, "memory", pj_string (j), 0);
+	pj_free (j);
+
+	ht_up_foreach (memory, save_memorylist, db);
+	return true;
+}
+
+static bool save_checkpoints(Sdb *db, RVector *checkpoints) {
+	size_t i;
+	RDebugCheckpoint *chkpt;
+	RDebugSnap *snap;
+	RListIter *iter;
+	PJ *j = pj_new ();
+	if (!j) {
+		return false;
+	}
+
+	pj_o (j);
+	r_vector_foreach (checkpoints, chkpt) {
+		// Append cnum to chkptlist sdb_array
+		// chkptlist=<cnum>,<cnum>,...
+		sdb_array_add (db, "chkptlist", sdb_fmt ("0x%"PFMT64x, chkpt->cnum), 0);
+
+		pj_ko (j, sdb_fmt ("0x%"PFMT64x, chkpt->cnum));
+
+		// Serialize RRegArena to json
+		// registers: {"<RRegisterType>": <RRegArena>, ...}
+		pj_ko (j, "registers");
+		for (i = 0; i < R_REG_TYPE_LAST; i++) {
+			iter = chkpt->reg[i];
+			RRegArena *arena = iter->data;
+			pj_ko (j, sdb_fmt ("%d", i));
+			if (arena->bytes) {
+				char *ebytes = sdb_encode ((const void *)arena->bytes, arena->size);
+				pj_ks (j, "bytes", ebytes);
+				free (ebytes);
+				pj_kn (j, "size", arena->size);
+			} else {
+				pj_kn (j, "size", 0);
+			}
+			pj_end (j);
+		}
+		pj_end (j);
+
+		// Serialize RDebugSnap to json
+		// snaps: {"size":<size_t>, "a":[<RDebugSnap>]},
+		pj_ko (j, "snaps");
+		pj_kn (j, "size", r_list_length (chkpt->snaps));
+		pj_ka (j, "a");
+		r_list_foreach (chkpt->snaps, iter, snap) {
+			pj_o (j);
+			pj_ks (j, "name", snap->name);
+			pj_kn (j, "addr", snap->addr);
+			pj_kn (j, "addr_end", snap->addr_end);
+			pj_kn (j, "size", snap->size);
+			char *edata = sdb_encode ((const void *)snap->data, snap->size);
+			pj_ks (j, "data", edata);
+			free (edata);
+			pj_kn (j, "perm", snap->perm);
+			pj_kn (j, "user", snap->user);
+			pj_kb (j, "shared", snap->shared);
+			pj_end (j);
+		}
+		pj_end (j);
+		pj_end (j);
+		pj_end (j);
+	}
+	pj_end (j);
+	sdb_set (db, "checkpoints", pj_string (j), 0);
+	pj_free (j);
+	return true;
+}
+
+/*
+ * SDB Format:
+ *
+ * /
+ *   maxcnum=<maxcnum>
+ *
+ *   reglist=<addr>,<addr>,...
+ *   registers={"<addr>": {"size":<size_t>, "a":[<RDebugChangeReg>]}, ...}
+ *
+ *   memorylist=<addr>,<addr>,...
+ *   memory={"<addr>": {"size":<size_t>, "a":[<RDebugChangeMem>]}, ...}
+ *
+ *   chkptlist=<cnum>,<cnum>,...
+ *   checkpoints={
+ *     "<cnum>": {
+ *       registers: {"<RRegisterType>": <RRegArena>, ...},
+ *       snaps: {"size":<size_t>, "a":[<RDebugSnap>]},
+ *     },
```

Comment:
Having such a big dictionary in a single sdb value can become an issue. SDB by design has a maximum value size of 16MB (please correct me if I'm wrong @trufae) and I would expect these checkpoints to commonly exceed that. The same might be the case for `registers` and `memorylist`.
What I would do here is to use one `sdb_ns` for each of these lists and use the sdb structure of the sub-namespaces as the dict. Instead of `chkptlist` you can then just list all the keys of the respective namespace.

Reply:
> Having such a big dictionary in a single sdb value can become an issue. SDB by design has a maximum value size of 16MB (please correct me if I'm wrong @trufae) and I would expect these checkpoints to commonly exceed that. The same might be the case for registers and memorylist.

You are right. I should just dump the memory snapshots into their own files rather than saving them in base64 and store them in `sdb`. Now, I see the reason why `Rangstr` is `unsigned short` which could be the max limit you mentioned. 

> What I would do here is to use one sdb_ns for each of these lists and use the sdb structure of the sub-namespaces as the dict. Instead of chkptlist you can then just list all the keys of the respective namespace.

That sounds better and I will just store them separately into its own db like how `r2db` does. I could wait for `r2db`, so I do not have to re-implement  the sdb archiving functions.

Reply:
It still makes sense to extend Rangstr since `unsigned short` has a max of 64k instead of 16MB. I think it depends if the memory snapshot base64s themselves can exceed 16MB whether it makes sense to dump them into separate binary files outside sdb. If there is only one per sdb key it might work out.

Reply:
Dumping them into their own binaries could save time for encoding and decoding them in base64. Besides, the only advantage of storing them in sdb is having everything in one place. I will just keep them in sdb for now.

Reply:
For now, I will just save each of them into its own `sdb` files. It would be great to have a function such as `sdb_archive_save` in `r2db` to store everything in one archive. Are you planning to add that in future?